src/Network/Transport/TCP.hs

```diff
@@ -1450,7 +1452,7 @@ schedule theirEndPoint act = do
 -- 'runScheduledAction' or by another).
 runScheduledAction :: EndPointPair -> Action a -> IO a
 runScheduledAction (ourEndPoint, theirEndPoint) mvar = do
-    join $ readChan (remoteScheduled theirEndPoint)
+    join $ uninterruptibleMask_ $ readChan (remoteScheduled theirEndPoint)
```

Comment:
> If the chan is always not empty, then a simple mask is sufficient.

I don't thinks that's correct. 'mask_ $ readChan ...' can be interrupted when the channel is not empty according to the implementation of 'readChan'.

I have expanded a bit the description in https://cloud-haskell.atlassian.net/browse/DP-109. If you think there is a better solution or that the problem is not so, it is better to mention it there.